### PR TITLE
[CDAP-14282] Adds UI testing infrastructure & initial tests

### PR DESCRIPTION
**Pull request by tonybach**
JIRA: https://issues.cask.co/browse/CDAP-14282
UI builds: https://builds.cask.co/browse/CDAP-UDUT145
IT builds: https://builds.cask.co/browse/IT-UPD20

To run tests locally:
- all tests in parallel: `yarn test`.
- snapshot/integration tests only: `yarn jest`.
- run end-to-end (Cypress) tests only: `yarn cypress`.
- run end-to-end (Cypress) tests, with steps rendered in browser: `yarn cypress-open`, then click `Run all specs`

### CODE REVIEW

File: cdap-ui/app/cdap/components/Tags/__tests__/Tags.test.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    await wait(() => !queryByTestId('loading-icon')); // wait until loading icon is gone, which means async call is done
+
+    const plusButtonElem = getByTestId('tag-plus-button');
+    fireEvent.click(plusButtonElem, {container}); // click on plus button to show input box
+    await waitForElement(() => getByTestId('tag-input'));
+
+    const inputElem = getByTestId('tag-input');
+    expect(inputElem).toHaveTextContent('');
+    fireEvent.change(inputElem, {target: {value: TEST_TAG}});
+    Mousetrap.trigger('return'); // enter value, and then press Enter to add tag
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Tags/__tests__/Tags.test.js]: is it possible to just use `fireEvent` for this instead of using Mousetrap?

Comment (tonybach) [reply cdap-ui/app/cdap/components/Tags/__tests__/Tags.test.js]: Yes we should be able to. Will update the PR

**Review by ajainarayanan — COMMENTED** [review]
@tonybach Overall the changes look good. Couple of things,

1. Could we have fixed version of libraries that we are adding
2. In the cypress test what happens if the NUX is enabled?
3. Referred to the design doc but I couldn't find a clear definition of when to add a snapshot test to which component. Is the assumption of adding a snapshot test is only for dumb components like buttons or popovers etc.,? Was there any specific reason for adding snapshot tests for Tags?
4. Does this PR also address running cypress tests in Bamboo or is it only unit and integration tests as of now?

**Comment by tonybach** [comment]
@ajainarayanan 
1. Yes have added fixed versions.
2. Since in the tests so far we're not explicitly testing the behavior of the NUX, before visiting the homepage I set `showWelcome` to `false` in `sessionStorage`. This will hide the NUX: https://github.com/caskdata/cdap/pull/10817/files#diff-a383fc4a0f57f2016c83c830d81af0a5R71
3. We should have a snapshot test for every single React component, so I added the snapshot tests to Tags as well.
4. I set up integration with Bamboo for both:
- For unit/integration tests, I added a step to the pom.xml, so that the tests are always run during the UI builds (UDUT/URUT).
- For e2e/cypress tests, I set up this build https://builds.cask.co/browse/IT-UPD which should run the tests every night, based off whatever is in the `develop` branch.

File: cdap-ui/cypress/integration/pipeline.spec.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    cy.contains('Pipeline config').click();
+    cy.get('.label-with-toggle')
+      .contains('Instrumentation')
+      .parent()
+      .as('instrumentationDiv');
+    cy.get('@instrumentationDiv').contains('Off');
+    cy.get('[data-testid=close-modeless]').click();
+
+    // Delete the pipeline to clean up
+    cy.get('.pipeline-actions-popper').click();
```
Comment (elfenheart) [inline cdap-ui/cypress/integration/pipeline.spec.js]: if the intent here is to clean up, i think there should be a method `afterEach` that will do the cleanup process?

If the intent is to test the deletion process, then it should be under a different test case, I think.

Comment (tonybach) [reply cdap-ui/cypress/integration/pipeline.spec.js]: You're right, makes sense to do that. I've moved the cleanup block to be in the `afterEach` hook.

File: cdap-ui/cypress/integration/pipeline.spec.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  afterEach(function() {
+    // Delete the pipeline to clean up
+    cy.get('.pipeline-actions-popper').click();
+    cy.get('[data-testid=delete-pipeline]').click();
+    cy.get('[data-testid=confirmation-modal]')
+      .find('.btn-primary')
+      .click();
+
+    // Assert pipeline no longer exists
+    cy.contains('a', TEST_PIPELINE_NAME, { timeout: 10000 }).should('not.exist');
```
Comment (elfenheart) [inline cdap-ui/cypress/integration/pipeline.spec.js]: should remove this... there's shouldn't be an assertion on a setup/cleanup process

**Review by elfenheart — DISMISSED** [review]
👍 LGTM

**Comment by tonybach** [comment]
Latest builds are green: https://builds.cask.co/browse/CDAP-UDUT145-14, https://builds.cask.co/browse/IT-UPD20-66.

**Review by elfenheart — APPROVED** [review]

**Pull request merged**